src/main/java/apoc/meta/tablesforlabels/PropertyContainerProfile.java

```diff
@@ -0,0 +1,98 @@
+package apoc.meta.tablesforlabels;
+
+import org.neo4j.graphdb.Label;
+import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.PropertyContainer;
+import org.neo4j.graphdb.schema.ConstraintDefinition;
+
+import java.util.*;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A profile of a particular ordered label set (or relationship type) has a set of possible properties that can exist, and
+ * stats about those properties.
+ */
+public class PropertyContainerProfile {
+    public long observations;
+    Map<String, PropertyTracker> profile;
+
+    public PropertyContainerProfile() {
+        observations = 0;
+        profile = new HashMap<>(3);
+    }
+
+    public Set<String> propertyNames() { return profile.keySet(); }
+    public PropertyTracker trackerFor(String propName) { return profile.get(propName); }
+
+    public void observe(PropertyContainer n, Iterable<ConstraintDefinition> constraints, boolean isNode, Map<String, Iterable<ConstraintDefinition>> relConstraints, Iterable<Label> relStartNode, Iterable<Label> relEndNode) {
+        observations++;
+
+        for (String propName : n.getPropertyKeys()) {
+            PropertyTracker tracker;
+
+            if (profile.containsKey(propName)) {
+                tracker = profile.get(propName);
+            } else {
+                tracker = new PropertyTracker(propName);
+                profile.put(propName, tracker);
+            }
+
+            tracker.addObservation(n.getProperty(propName));
+
+            if (isNode) {
+
+                // Check for node constraints
+
+                tracker.mandatory = false;
+                for (ConstraintDefinition cd : constraints) {
```

Comment:
Each time a property container (i.e. an individual node is observed) we're looking at the constraints.   They'll never change, they should be examined once upfront and not every time we look at a node, because this will add unnecessary processing time.

Additionally -- a thing to be super careful of is that mandatory = true **only when** there is an **existence constraint**, not just any constraint.  If for example there is a uniqueness constraint, this will cause mandatory=true, which is wrong.   Here's an example of why:

```
CREATE CONSTRAINT ON (f:Foo) ASSERT f.x is unique;
create (:Foo { x: null });
create (:Foo { x: null });
```

That works because null is never equal to itself in any database, and so they are unique.  But they don't exist.  

So set mandatory=true only for existance constraints, and do it only once.

Reply:
Existence check was added a while ago, so this one is done!